### PR TITLE
box: filter out tmp and remote txs when aborting all txs for DDL

### LINKED ISSUE #9720: Temporary spaces abort all DDL even without MVCC
The reproducer requires 2 instances:
```Lua
--
-- Instance 1
--
-- Step 1
--
box.cfg{
    listen = 3301,
    replication = {3301, 3302},
}
box.once('grant guest', function() 
    box.schema.user.grant('guest', 'read,write,execute', 'universe')
end)

--
-- Step 3
--
format = {{name = 'id', type = 'unsigned'}}
space = box.schema.create_space('test', {format = format})
_ = space:create_index('pk')
do
    box.begin()
    for i = 1, 1000 do
        space:replace{i, i}
    end
    box.commit()
end

--
-- Step 5
--
table.insert(format, {name = 'val', type = 'unsigned'})
space:format(format)
```
and
```Lua
--
-- Instance 2
--
-- Step 2
--
box.cfg{
    listen = 3302,
    replication = {3301, 3302},
}
box.once('grant guest', function() 
    box.schema.user.grant('guest', 'read,write,execute', 'universe')
end)

--
-- Step 4
--
box.error.injection.set('ERRINJ_CHECK_FORMAT_DELAY', true)

--
-- Step 6
--
space = box.schema.space.create('TMP', { type = 'temporary' })
space:drop()
```

The result is that the step-6 will abort the long DDL txn in the applier, combing from the instance-1, even though those 2 DDL txns are not related anyhow. The temporary spaces don't even yield or store any actual data on disk.

This hits the users who are trying to use the temporary spaces as short living objects, creating and deleting them frequently.

The bug seems to be appearing since this commit: https://github.com/tarantool/tarantool/commit/8f4be3227635ef606581dda7e85f254a96347c0e.

### PR DESCRIPTION
**Pull request by CuriousGeorgiy**
This patch series filters out temporary and remote transactions when aborting all transactions for DDL, since they cannot conflict each other by definition.

Closes #9720

### CODE REVIEW

**Review by andreyaksenov — APPROVED** [review]

File: src/box/txn.h [hunk 1]
```diff
@@ -885,6 +889,12 @@ txn_is_nop(const struct txn *txn)
 	return txn->n_new_rows + txn->n_applier_rows == 0;
 }
 
+static inline bool
+txn_is_remote(const struct txn *txn)
```
Comment (drewdzzz) [inline src/box/txn.h:893]: Nit: maybe call it `txn_is_fully_remote`? (analogy with `txn_is_fully_local`).

File: src/box/memtx_tx.c [hunk 2]
```diff
@@ -683,6 +683,17 @@ memtx_tx_acquire_ddl(struct txn *tx)
 	(void) txn_can_yield(tx, false);
 }
 
+/**
+ * Nop and remote transactions cannot possibly conflict each other by
+ * definition, so we can filter them out when aborting transactions for DDL.
+ */
+static bool
+memtx_tx_filter_nop_and_remote_txs(struct txn *txn1, struct txn *txn2)
```
Comment (drewdzzz) [inline src/box/memtx_tx.c:691]: Actually, `nop` shouldn't conflict anything since it means no-op. But our `nop` is not no-op - it can be a temporary statement, and it's a bit confusing here. Let's, at least, call this function like `memtx_tx_filter_temporary_and_remote_txs` with elaboration in commentary so that it's clear that `nop` is not always a no-op.

Comment (drewdzzz) [reply src/box/memtx_tx.c:691]: Also, cannot we use `txn_is_fully_local()` here instead of `txn_is_nop()`? Any temporary transaction is fully local, and it seems that any local transaction cannot conflict with purely remote one.

File: src/box/txn.c [hunk 3]
```diff
@@ -653,6 +653,8 @@ txn_commit_stmt(struct txn *txn, struct request *request)
 		assert(stmt->row != NULL);
 		txn_update_row_counts(txn, stmt, 1);
 	}
+	if (stmt->space != NULL && space_is_data_temporary(stmt->space))
```
Comment (drewdzzz) [inline src/box/txn.c:656]: Isn't it complement of the condition above? It could be just `else` clause.

**Review by drewdzzz — CHANGES_REQUESTED** [review]

File: src/box/txn.h [hunk 4]
```diff
@@ -525,6 +525,10 @@ struct txn {
 	 * Nesting level of space on_replace triggers for current txn.
 	 */
 	int space_on_replace_triggers_depth;
+	/**
+	 * Whether the transaction has temporary statements.
+	 */
+	bool has_temporary_stmts;
```
Comment (drewdzzz) [inline src/box/txn.h:531]: Let's elaborate in description why we need the flag at all (rows of temporary statements are erased).

File: src/box/txn.h (hunk 1 above)
Comment (CuriousGeorgiy) [reply src/box/txn.h:893]: Okay, done.

File: src/box/txn.c (hunk 3 above)
Comment (CuriousGeorgiy) [reply src/box/txn.c:656]: No, the condition above has an `or`. We shouldn't set the flag if the statement doesn't have a space assigned to it.

Comment (CuriousGeorgiy) [reply src/box/txn.c:656]: True, thanks!

File: src/box/memtx_tx.c (hunk 2 above)
Comment (CuriousGeorgiy) [reply src/box/memtx_tx.c:691]: > Any temporary transaction is fully local, and it seems that any local transaction cannot conflict with purely remote one.

A transaction on data-temporary spaces (for which the space metadata is replicated) can be conflicting, so we need `txn_is_nop` here.

File: src/box/txn.h (hunk 4 above)
Comment (CuriousGeorgiy) [reply src/box/txn.h:531]: Done.

File: src/box/memtx_tx.c (hunk 2 above)
Comment (drewdzzz) [reply src/box/memtx_tx.c]: Oh, thanks, I see. But `txn_is_nop` is not sufficient then. If a transaction did a replace in a data-temporary space, it is `nop` but it actually can conflict with remote one since it can do DDL on data-temporary spaces.

Comment (CuriousGeorgiy) [reply src/box/memtx_tx.c]: No, AFAIC, it is enough. A transaction contaning DDL on data-temporary spaces would have `n_new_rows > 0`, since the DDL statement would get written to the WAL (see `txn_is_fully_local`).

Comment (drewdzzz) [reply src/box/memtx_tx.c]: AFAICS, we [update](https://github.com/tarantool/tarantool/blob/e01c20edc31b65e020ac57e3f62a7427b6e27d53/src/box/txn.c#L650) row counts only for non-data-temporary statements.

Comment (CuriousGeorgiy) [reply src/box/memtx_tx.c]: > So, we can have local DML transaction on data-temporary space, which is nop, and we can have transaction in applier doing DDL on this data-temporary space.

If we have a transaction in the applier doing DDL, it acquires DDL, so it is forbideen to yield — even with MVCC enabled:
https://github.com/tarantool/tarantool/blob/e01c20edc31b65e020ac57e3f62a7427b6e27d53/src/box/memtx_tx.c#L679-L684

Unless it is a space format change or a new index creation, for which yielding is safe by design.

Comment (drewdzzz) [reply src/box/memtx_tx.c]: Consider the scenario:
```
TX1: BEGIN
TX1: DML on data-temporary space
APPLIER: DDL on data-temporary space
TX1: COMMIT
```
Here applier should abort TX1 since they are operating the same space, but it doesn't work in this way in your patch because TX1 is actually `nop` (it has no rows) and APPLIER is `fully_remote`.

Comment (CuriousGeorgiy) [reply src/box/memtx_tx.c]: You are missing the point: the applier transaction cannot yield here, otherwise it will be aborted by yield.

File: src/box/txn.c [hunk 5]
```diff
@@ -652,6 +652,8 @@ txn_commit_stmt(struct txn *txn, struct request *request)
 			goto fail;
 		assert(stmt->row != NULL);
 		txn_update_row_counts(txn, stmt, 1);
+	} else {
+		txn->has_temporary_stmts = true;
```
Comment (Gerold103) [inline src/box/txn.c:656]: That unfortunately won't work well if the statement is rolled back by a savepoint-rollback. The flag would remain set but inaccurate. I would suggest to rather walk the statements in `txn_is_fully_remote()` to check if any of them are temporary. It is linear complexity, but is needed only for true DDL (i.e. not on temporary spaces), which can only be rare. Given that it aborts all concurrent txns (excluding applier and temporary spaces).

Also the flag isn't initialized now in `txn_begin`. Can as well become `true` randomly and start failing. And for flags anyway we use `enum txn_flag`. But I suggest not to add it at all really. It is deductible from the statements and is needed only on rollback (perf isn't important).

File: test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    cg.server2:exec(function()
+        box.error.injection.set('ERRINJ_CHECK_FORMAT_DELAY', true)
+    end)
+    cg.server1:exec(function()
+        box.space.s:format{'i', 'unsigned'}
+    end)
+    cg.server2:exec(function()
+        box.schema.space.create('tmp', {type = 'temporary'})
+        box.error.injection.set('ERRINJ_CHECK_FORMAT_DELAY', false)
+    end)
```
Comment (Gerold103) [inline test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua:61]: Does it fail without the patch? It looks like the last exec is likely to be executed before the second one is replicated. So the format coming to `server2` won't even see `ERRINJ_CHECK_FORMAT_DELAY = true` and won't try to abort anything.

File: test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+    }
+    cg.replica_set:start()
+    cg.server1:exec(function()
+        local s = box.schema.create_space('s')
+        s:create_index('p')
+        box.begin()
+        for i = 1, 1000 do
+            s:replace{i, i}
+        end
+        box.commit()
```
Comment (Gerold103) [inline test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua:34]: Why do you need so many tuples in the space? Same below in `test_remote_ddl()`.

File: test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+    t.skip_if(not cg.params.memtx_use_mvcc)
+
+    cg.server2:exec(function()
+        local tmp = box.schema.space.create('tmp-replace', {type = 'temporary'})
+        tmp:create_index('p')
+
+        box.space.s:on_replace(function()
+            tmp:replace{0}
+            rawset(_G, 'executed_on_replace', true)
+            while true do require('fiber').yield() end
```
Comment (Gerold103) [inline test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua:82]: That is going to be a busy loop with almost 100% CPU usage. Lets better make it sleep(some big value) or `.stall()` (sleep until wakeup or cancel).

File: test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+    end)
+    cg.server2:exec(function()
+        t.helpers.retrying({timeout = 60}, function()
+            t.assert(_G.executed_on_replace)
+        end)
+        box.schema.space.create('tmp', {type = 'temporary'})
+    end)
+    local msg = "Transaction committing DDL (id=%d+) has aborted " ..
+                "another TX (id=%d+)"
+    cg.server2:grep_log(msg)
```
Comment (Gerold103) [inline test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua:96]: `grep_log` won't fail on its own, you need to check it with `t.assert`.

**Review by Gerold103 — COMMENTED** [review]
Thanks, good patch 💪.

File: src/box/memtx_tx.c (hunk 2 above)
Comment (CuriousGeorgiy) [reply src/box/memtx_tx.c]: Sorry for my misconception, and thanks for the catch, fixed: added another flag to track data-temporary statements, and added a test case for this scenario.

File: src/box/txn.c (hunk 5 above)
Comment (CuriousGeorgiy) [reply src/box/txn.c:656]: > I would suggest to rather walk the statements in `txn_is_fully_remote()` to check if any of them are temporary.

I don't understand how you suggest to check if they are temporary. Temporary and nop statements are indistinguishable from this point in the code. Only if we add a `temporary` flag to `txn_stmt`.

> That unfortunately won't work well if the statement is rolled back by a savepoint-rollback. The flag would remain set but inaccurate.

Is this a common case? If not, I would suggest leaving it as is, since we are trying to fix a common source of pain for Picodata user workloads here.

> Also the flag isn't initialized now in `txn_begin`.

Fixed, thanks.

> And for flags anyway we use `enum txn_flag`.

Not always, for instance we have a `is_schema_changed` flag. I think this is a temporary solution, and it should also be just a separate flag.

File: test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua (hunk 7 above)
Comment (CuriousGeorgiy) [reply test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua:34]: Because otherwise I won't get to the error injection:
https://github.com/tarantool/tarantool/blob/09bfdd473ffd23169dbe8d4e5d0510e8c70b11ec/src/box/memtx_space.c#L52-L65

File: test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua (hunk 6 above)
Comment (CuriousGeorgiy) [reply test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua:61]: Yes, it fails. I see your point, but how do you suggest setting up a synchronization point here?

I tried to make the `_space` synchronous and wait for `box.info.synchro.queue.busy` on `server2`, but the replication connection starts to timeout then:
```console
2024-08-03 12:15:11.398 [42760] relay/unix/:(socket)/101/main coio.c:300 E> TimedOut: timed out
2024-08-03 12:15:11.398 [42760] relay/unix/:(socket)/101/main I> exiting the relay loop
```

File: test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua (hunk 9 above)
Comment (CuriousGeorgiy) [reply test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua]: Thanks, fixed!

File: test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua (hunk 8 above)
Comment (CuriousGeorgiy) [reply test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua]: `stall` is not public, AFAIC:
https://github.com/tarantool/tarantool/blob/09bfdd473ffd23169dbe8d4e5d0510e8c70b11ec/src/lua/fiber.lua#L75-L80

Thanks for the catch, changed it to `t.helpers.retrying`.

File: src/box/txn.c (hunk 5 above)
Comment (CuriousGeorgiy) [reply src/box/txn.c]: Fixed by applying your suggestion, thanks for the catch! Added test cases for savepoint-rollback.

File: src/box/txn.h [hunk 10]
```diff
@@ -893,6 +893,38 @@ txn_is_nop(const struct txn *txn)
 	return txn->n_new_rows + txn->n_applier_rows == 0;
 }
 
+static inline bool
+txn_is_fully_temporary(struct txn *txn)
```
Comment (Gerold103) [inline src/box/txn.h:897]: Nit: the txn can be `const` here, no? Same for `txn_is_fully_remote()`.

File: src/box/txn.h [hunk 11]
```diff
@@ -893,6 +893,38 @@ txn_is_nop(const struct txn *txn)
 	return txn->n_new_rows + txn->n_applier_rows == 0;
 }
 
+static inline bool
```
Comment (Gerold103) [inline src/box/txn.h:896]: Nit: the function is relatively big, I would say it doesn't have to be `inline`. Only increases the compilation time this way. Same for `txn_is_fully_remote()`.

File: src/box/txn.h [hunk 12]
```diff
@@ -893,6 +893,38 @@ txn_is_nop(const struct txn *txn)
...
+txn_is_fully_temporary(struct txn *txn)
+{
+	if (!txn_is_nop(txn))
+		return false;
+	struct txn_stmt *stmt;
+	stailq_foreach_entry(stmt, &txn->stmts, next) {
+		if (stmt->space != NULL && !space_is_temporary(stmt->space) &&
+		    space_is_data_temporary(stmt->space))
```
Comment (Gerold103) [inline src/box/txn.h:904]: Nit: hm, that looks weird. Perhaps it would be better to check here literally `space->opts.type == SPACE_TYPE_DATA_TEMPORARY`. Instead of these 2 checks.

File: src/box/txn.h [hunk 13]
```diff
@@ -893,6 +893,38 @@ txn_is_nop(const struct txn *txn)
...
+{
+	if (txn->n_new_rows != 0)
+		return false;
+	struct txn_stmt *stmt;
+	stailq_foreach_entry(stmt, &txn->stmts, next) {
+		/*
+		 * Allow DDL on data-temporary spaces, since we allow only fully
+		 * temporary transactions to continue.
+		 */
+		if (stmt->row == NULL && stmt->space != NULL &&
```
Comment (Gerold103) [inline src/box/txn.h:921]: Nit: the function is just a checker, looks strange that it mentions DDL. I guess this comments rather belongs to the place, where this function is used?

**Review by Gerold103 — APPROVED** [review]
Thanks for the patch, I like how all the corner cases are covered in the tests 💪.

File: test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua (hunk 6 above)
Comment (Gerold103) [reply test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua:61]: Easy but not reliable way - just add here a `fiber.sleep(0.01)` to make it more likely to fail without the patch. Hard, but more correct way - introduce a errinj like `ERRINJ_WAL_DELAY_COUNTDOWN` or other `*_COUNTDOWN`s. An injection which gets set on Nth check. Such as `ERRINJ_CHECK_FORMAT_DELAY_COUNTDOWN`. And make it set `ERRINJ_CHECK_FORMAT_DELAY`. Then you can set countdown = 0, and wait until delay = true.

File: src/box/txn.h (hunk 10 above)
Comment (CuriousGeorgiy) [reply src/box/txn.h:897]: No, unfortunately, using the `stailq` API requires the original struct to be non-const:
```console
/Users/glebedev/Work/tarantool/src/box/txn.c:671:2: error: passing 'const struct stailq *' to parameter of type 'struct stailq *' discards qualifiers [-Werror,-Wincompatible-pointer-types-discards-qualifiers]
        stailq_foreach_entry(stmt, &txn->stmts, next) {

```

File: src/box/txn.h (hunk 12 above)
Comment (CuriousGeorgiy) [reply src/box/txn.h:904]: Fixed.

File: src/box/txn.h (hunk 13 above)
Comment (CuriousGeorgiy) [reply src/box/txn.h:921]: This function determines the filtering condition. This comment explains why we take filter out only statements with `stmt->row == NULL`.

File: test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua (hunk 6 above)
Comment (CuriousGeorgiy) [reply test/replication-luatest/gh_9720_tmp_spaces_ddl_abort_applier_txs_test.lua]: Thanks, added a `ERRINJ_CHECK_FORMAT_DELAY_COUNTDOWN` injection.

File: src/box/txn.h (hunk 13 above)
Comment (drewdzzz) [reply src/box/txn.h]: AFAIU, you expect that replace in a data-temporary space has `stmt->row == NULL` and DDL there has `stmt->row != NULL`. Well, any DDL is actually a replace in space `_space` which is not temporary at all, so it still would be allowed without this condition. Please, drop it.

**Review by drewdzzz — APPROVED** [review]
Thanks for the patch, LGTM!
Please, address my comment in the last Vlad's discussion before merging.

File: src/box/txn.h (hunk 13 above)
Comment (CuriousGeorgiy) [reply src/box/txn.h]: @drewdzzz Thanks, moved the `stmt->row == NULL` to an assert and moved the comment out of the loop.

File: src/box/memtx_space.c [hunk 14]
```diff
@@ -1073,6 +1073,13 @@ memtx_space_check_format(struct space *space, struct tuple_format *format)
 
 		ERROR_INJECT_YIELD(ERRINJ_CHECK_FORMAT_DELAY);
 
+		ERROR_INJECT_COUNTDOWN(ERRINJ_CHECK_FORMAT_DELAY_COUNTDOWN, {
+			struct errinj *e =
+				errinj(ERRINJ_CHECK_FORMAT_DELAY, ERRINJ_BOOL);
+			e->bparam = true;
+			ERROR_INJECT_YIELD(ERRINJ_CHECK_FORMAT_DELAY);
+		});
```
Comment (Gerold103) [inline src/box/memtx_space.c:1081]: If you switch these 2 ERROR_INJECT calls, then you can drop the one that you have on line 1080.

File: src/box/vinyl.c [hunk 15]
```diff
@@ -1150,6 +1150,12 @@ vinyl_space_check_format(struct space *space, struct tuple_format *format)
 		if (++loops % VY_YIELD_LOOPS == 0)
 			fiber_sleep(0);
 		ERROR_INJECT_YIELD(ERRINJ_CHECK_FORMAT_DELAY);
+		ERROR_INJECT_COUNTDOWN(ERRINJ_CHECK_FORMAT_DELAY_COUNTDOWN, {
+			struct errinj *e =
+				errinj(ERRINJ_CHECK_FORMAT_DELAY, ERRINJ_BOOL);
+			e->bparam = true;
+			ERROR_INJECT_YIELD(ERRINJ_CHECK_FORMAT_DELAY);
+		});
```
Comment (Gerold103) [inline src/box/vinyl.c:1158]: Same here. Or do I miss why it wouldn't work?

**Review by Gerold103 — COMMENTED** [review]
Thanks for the fixes!

File: src/box/memtx_space.c (hunk 14 above)
Comment (CuriousGeorgiy) [reply src/box/memtx_space.c:1081]: How will this work, if one wants to enable `ERRINJ_CHECK_FORMAT_DELAY` independently of `ERRINJ_CHECK_FORMAT_DELAY_COUNTDOWN`? For instance, if the latter is set to `-1` (default).

Also refer to `ERRINJ_WAL_DELAY{_COUNTDOWN}`?

File: src/box/vinyl.c (hunk 15 above)
Comment (CuriousGeorgiy) [reply src/box/vinyl.c:1158]: `ERROR_INJECT_COUTNDOWN` executes the code only when it has the `0` value set. The default value is `-1`, so `ERRINJ_CHECK_FORMAT_DELAY` won’t work.

Or maybe I don’t understand what you mean by
> If you switch these 2 ERROR_INJECT calls

File: src/box/memtx_tx.c (hunk 2 above)
Comment (Gerold103) [reply src/box/memtx_tx.c]: Resolving since this is approved anyway.

File: src/box/memtx_space.c (hunk 14 above)
Comment (Gerold103) [reply src/box/memtx_space.c:1081]: I guess we will ignore this for now.

File: src/box/vinyl.c (hunk 15 above)
Comment (Gerold103) [reply src/box/vinyl.c:1158]: I guess we will ignore this for now.

File: src/box/memtx_space.c (hunk 14 above)
Comment (CuriousGeorgiy) [reply src/box/memtx_space.c:1081]: Sorry, I forgot to push the fix.

File: src/box/vinyl.c (hunk 15 above)
Comment (CuriousGeorgiy) [reply src/box/vinyl.c:1158]: Sorry, I forgot to push the fix.
